### PR TITLE
kubeadm: aggregate all the errors when the shared certs are validated

### PR DESCRIPTION
**Pull request by chendave**
Instead of the individual error and return, it's better to aggregate all the errors so that we can fix them all at once.

Take the chance to fix some comments, since kubeadm are not checking that the certs are equal across controlplane.

Signed-off-by: Dave Chen <dave.chen@arm.com>

#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
thanks, i will have a look tomorrow.

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ import (
 	"k8s.io/client-go/util/keyutil"
 	"k8s.io/klog/v2"
 
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs.go:32]: this should be in the group above:

```
"k8s.io/client-go/util/keyutil"
"k8s.io/klog/v2"
```

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 2]
```diff
@@ -304,30 +305,32 @@ type certKeyLocation struct {
...
 	// in case of local/stacked etcd
 	if cfg.Etcd.External == nil {
 		if err := validateCACertAndKey(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.EtcdCACertAndKeyBaseName, "", "etcd CA"}); err != nil {
-			return false, err
+			errs = append(errs, err)
 		}
 	}
-
+	if len(errs) != 0 {
+		return false, utilerrors.NewAggregate(errs)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs.go:332]: this is worth updating the unit tests to check how many errors were returned exactly:
https://github.com/kubernetes/kubernetes/blob/e363bbdddb17100e6cc41a36c83caac49ea4c3c7/cmd/kubeadm/app/phases/certs/certs_test.go#L378

`expectedError bool` -> `expectedErrors int`

`len(err) != len(expectedErrors) ----> unit test failed, print err`

**Comment by neolit123** [comment]
/triage accepted
/priority backlog

LGTM, other than the two comments above.

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 2 above)
Comment (chendave) [reply cmd/kubeadm/app/phases/certs/certs.go:332]: Just found that testcase "missing etcd/ca.crt" is not wrote correctly,  it is just a copy of "missing front-proxy.crt".  😄

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 1 above)
Comment (chendave) [reply cmd/kubeadm/app/phases/certs/certs.go]: Done

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 2 above)
Comment (chendave) [reply cmd/kubeadm/app/phases/certs/certs.go:332]: Done

File: cmd/kubeadm/app/phases/certs/certs_test.go [hunk 3]
```diff
@@ -472,11 +484,12 @@ func TestSharedCertificateExists(t *testing.T) {
...
 			// executes create func
 			ret, err := SharedCertificateExists(cfg)
-
 			switch {
-			case !test.expectedError && err != nil:
-				t.Errorf("error SharedCertificateExists failed when not expected to fail: %v", err)
-			case test.expectedError && err == nil:
+			case err != nil:
+				if agg, ok := err.(utilerrors.Aggregate); ok && len(agg.Errors()) != test.expectedErrors {
+					t.Errorf("SharedCertificateExists didn't failed with expected error numbers")
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs_test.go:490]:
```suggestion
					t.Errorf("SharedCertificateExists didn't fail with the expected number of errors")
```

File: cmd/kubeadm/app/phases/certs/certs_test.go [hunk 4]
```diff
@@ -472,11 +484,12 @@ func TestSharedCertificateExists(t *testing.T) {
...
+				}
+			case err == nil && test.expectedErrors != 0:
 				t.Errorf("error SharedCertificateExists didn't failed when expected")
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs_test.go:493]:
```suggestion
				t.Errorf("error SharedCertificateExists didn't fail when expected")
```

^ there is also a typo in the old code

**Review by neolit123 — COMMENTED** [review]
found a couple of typos.
please keep the commits squashed to 1.

**Comment by neolit123** [comment]
/retitle kubeadm: aggregate all the errors when the shared certs are validated

File: cmd/kubeadm/app/phases/certs/certs_test.go (hunk 4 above)
Comment (chendave) [reply cmd/kubeadm/app/phases/certs/certs_test.go:493]: :)

File: cmd/kubeadm/app/phases/certs/certs_test.go (hunk 3 above)
Comment (chendave) [reply cmd/kubeadm/app/phases/certs/certs_test.go]: Good catch! Done

**Comment by chendave** [comment]
@neolit123 thanks for the review, could you pls take a look again?

**Review by neolit123 — COMMENTED** [review]
thanks
/lgtm
/approve

**Pull request merged**